Append tests failing with xarray 0.19.0

With xarray 0.19.0, the following two tests are failing:

```
DatasetWriterTest.test_append_overlapping_append_newer 
>       np.testing.assert_equal(expected, ds3.t.data)
E       AssertionError: 
E       Arrays are not equal
E       
E       (shapes (5,), (3,) mismatch)
E        x: array(['2001-01-01T00:00:00.000000000', '2001-01-02T00:00:00.000000000',
E              '2001-01-03T00:00:00.000000000', '2001-01-04T00:00:00.000000000',
E              '2001-02-05T00:00:00.000000000'], dtype='datetime64[ns]')
E        y: array(['2001-01-01T00:00:00.000000000', '2001-01-02T00:00:00.000000000',
E              '2001-01-03T00:00:00.000000000'], dtype='datetime64[ns]')

DatasetWriterTest.test_append_overlapping_retain 
E       AssertionError: 
E       Arrays are not equal
E       
E       (shapes (6,), (7,) mismatch)
E        x: array(['2001-01-01T00:00:00.000000000', '2001-01-02T00:00:00.000000000',
E              '2001-01-03T00:00:00.000000000', '2001-01-04T00:00:00.000000000',
E              '2001-01-05T00:00:00.000000000', '2001-01-06T00:00:00.000000000'],
E             dtype='datetime64[ns]')
E        y: array(['2001-01-01T00:00:00.000000000', '2001-01-02T00:00:00.000000000',
E              '2001-01-03T00:00:00.000000000', '2001-01-04T00:00:00.000000000',
E              '2001-01-05T00:00:00.000000000', '2001-01-05T00:00:00.000000000',
E              '2001-01-06T00:00:00.000000000'], dtype='datetime64[ns]')
```

The failures seem to be connected with the introduction of `consolidated=True` as the default for `xarray.open_zarr`.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,5 +1,9 @@
 ## nc2zarr Change History
 
+### Version 1.3.0 (in development)
+
+* Handle consolidated Zarrs correctly when appending data (fixes #47).
+
 ### Version 1.2.2
 
 * Fix incorrect version number.

--- a/nc2zarr/dataslice.py
+++ b/nc2zarr/dataslice.py
@@ -125,7 +125,14 @@ def update_slice(store: Union[str, MutableMapping],
 
     append_dim_var_names = []
     encoding = {}
-    with xr.open_zarr(store) as ds:
+
+    consolidated = True
+    try:
+        _ = zarr.open_consolidated(store)
+    except KeyError:
+        consolidated = False
+
+    with xr.open_zarr(store, consolidated=consolidated) as ds:
         for var_name in ds.variables:
             var = ds[var_name]
             if var.ndim >= 1 and dimension in var.dims:
@@ -167,3 +174,6 @@ def update_slice(store: Union[str, MutableMapping],
                     var_array[insert_index:-1, ...]
             # Replace slice
             var_array[insert_index, ...] = slice_array[0]
+
+    if consolidated:
+        zarr.consolidate_metadata(store)

--- a/nc2zarr/dataslice.py
+++ b/nc2zarr/dataslice.py
@@ -126,6 +126,13 @@ def update_slice(store: Union[str, MutableMapping],
     append_dim_var_names = []
     encoding = {}
 
+    # Neither Zarr nor xarray offer an explicit API function to check whether
+    # a Zarr is consolidated. Here we use the workaround of attempting to
+    # open as consolidated, and catching the resulting exception if this
+    # isn't possible. In the case of a consolidated Zarr, there is a slight
+    # inefficiency, since the consolidated metadata object is fetched twice
+    # (by Zarr and thereafter by xarray). See comments on PR #48 for
+    # discussion of possible optimizations.
     consolidated = True
     try:
         _ = zarr.open_consolidated(store)

--- a/tests/test_writer.py
+++ b/tests/test_writer.py
@@ -79,13 +79,13 @@ class DatasetWriterTest(unittest.TestCase, IOCollector):
         for i in range(3):
             ds = new_test_dataset(day=i + 1)
             writer.write_dataset(ds)
-        with xr.open_zarr('my.zarr') as ds:
+        with xr.open_zarr('my.zarr', consolidated=False) as ds:
             self.assertNotIn('history', ds.attrs)
             self.assertNotIn('source', ds.attrs)
             self.assertNotIn('time_coverage_start', ds.attrs)
             self.assertNotIn('time_coverage_end', ds.attrs)
         writer.finalize_dataset()
-        with xr.open_zarr('my.zarr') as ds:
+        with xr.open_zarr('my.zarr', consolidated=False) as ds:
             self.assertIn('history', ds.attrs)
             self.assertIn('source', ds.attrs)
             self.assertEqual('a.nc, b.nc', ds.attrs['source'])
@@ -96,16 +96,19 @@ class DatasetWriterTest(unittest.TestCase, IOCollector):
 
     def test_finalize_adjusts_metadata_with_time_bnds(self):
         self.add_path('my.zarr')
-        writer = DatasetWriter('my.zarr', output_append=True, output_adjust_metadata=True)
+        writer = DatasetWriter('my.zarr', output_append=True,
+                               output_adjust_metadata=True)
         for i in range(3):
             ds = new_test_dataset(day=i + 1, add_time_bnds=True)
             writer.write_dataset(ds)
         writer.finalize_dataset()
-        with xr.open_zarr('my.zarr') as ds:
+        with xr.open_zarr('my.zarr', consolidated=False) as ds:
             self.assertIn('time_coverage_start', ds.attrs)
-            self.assertEqual('2020-12-01 09:30:00', ds.attrs['time_coverage_start'])
+            self.assertEqual('2020-12-01 09:30:00',
+                             ds.attrs['time_coverage_start'])
             self.assertIn('time_coverage_end', ds.attrs)
-            self.assertEqual('2020-12-03 10:30:00', ds.attrs['time_coverage_end'])
+            self.assertEqual('2020-12-03 10:30:00',
+                             ds.attrs['time_coverage_end'])
 
     def test_finalize_updates_metadata(self):
         self.add_path('my.zarr')
@@ -115,10 +118,10 @@ class DatasetWriterTest(unittest.TestCase, IOCollector):
         for i in range(3):
             ds = new_test_dataset(day=i + 1)
             writer.write_dataset(ds)
-        with xr.open_zarr('my.zarr') as ds:
+        with xr.open_zarr('my.zarr', consolidated=False) as ds:
             self.assertNotIn('comment', ds.attrs)
         writer.finalize_dataset()
-        with xr.open_zarr('my.zarr') as ds:
+        with xr.open_zarr('my.zarr', consolidated=False) as ds:
             self.assertIn('comment', ds.attrs)
             self.assertEqual('This dataset is a test.', ds.attrs['comment'])
 
@@ -230,15 +233,16 @@ class DatasetWriterTest(unittest.TestCase, IOCollector):
 
     def test_local_postprocessor(self):
         self.add_path('my.zarr')
-        writer = DatasetWriter('my.zarr',
-                               output_overwrite=False,
-                               output_custom_postprocessor='tests.test_writer:my_postprocessor')
+        writer = DatasetWriter(
+            'my.zarr',
+            output_overwrite=False,
+            output_custom_postprocessor='tests.test_writer:my_postprocessor')
         ds = new_test_dataset(day=1)
         self.assertNotIn('crs', ds)
 
         writer.write_dataset(ds)
         self.assertTrue(os.path.isdir('my.zarr'))
-        with xr.open_zarr('my.zarr') as ds:
+        with xr.open_zarr('my.zarr', consolidated=False) as ds:
             self.assertIn('crs', ds)
 
     # noinspection PyMethodMayBeStatic
@@ -462,7 +466,7 @@ class DatasetWriterTest(unittest.TestCase, IOCollector):
 
     @classmethod
     def assertTimeSlicesOk(cls, dst_path, src_path_pat, n):
-        with xr.open_zarr(dst_path, decode_cf=False) as ds:
+        with xr.open_zarr(dst_path, decode_cf=False, consolidated=False) as ds:
             for i in range(0, n):
                 src_path = src_path_pat.format(i)
                 with xr.open_zarr(src_path, decode_cf=False) as src_dataset:

--- a/tests/test_writer.py
+++ b/tests/test_writer.py
@@ -377,22 +377,25 @@ class DatasetWriterTest(unittest.TestCase, IOCollector):
             w.write_dataset(ds2)
 
     def test_append_overlapping_append_newer(self):
-        dst_path = "my.zarr"
-        self.add_path(dst_path)
-        ds1, ds2 = new_append_test_datasets(
-            ["2001-01-01", "2001-01-02", "2001-01-03"],
-            ["2001-01-02", "2001-01-03", "2001-01-04", "2001-02-05"]
-        )
-        ds1.to_zarr(dst_path)
-        w = DatasetWriter(dst_path, output_append=True,
-                          output_append_dim="t",
-                          output_append_mode=AppendMode.newer)
-        w.write_dataset(ds2)
-        ds3 = xr.open_zarr(dst_path)
-        expected = np.array(["2001-01-01", "2001-01-02", "2001-01-03",
-                             "2001-01-04", "2001-02-05"],
-                            dtype="datetime64[ns]")
-        np.testing.assert_equal(expected, ds3.t.data)
+        for consolidated in False, True:
+            with self.subTest(consolidated=consolidated):
+                dst_path = "my.zarr"
+                self.add_path(dst_path)
+                ds1, ds2 = new_append_test_datasets(
+                    ["2001-01-01", "2001-01-02", "2001-01-03"],
+                    ["2001-01-02", "2001-01-03", "2001-01-04", "2001-02-05"]
+                )
+                ds1.to_zarr(dst_path, consolidated=consolidated)
+                w = DatasetWriter(dst_path, output_append=True,
+                                  output_append_dim="t",
+                                  output_append_mode=AppendMode.newer,
+                                  output_consolidated=consolidated)
+                w.write_dataset(ds2)
+                ds3 = xr.open_zarr(dst_path, consolidated=consolidated)
+                expected = np.array(["2001-01-01", "2001-01-02", "2001-01-03",
+                                     "2001-01-04", "2001-02-05"],
+                                    dtype="datetime64[ns]")
+                np.testing.assert_equal(expected, ds3.t.data)
 
     def test_append_non_increasing_append_newer(self):
         dst_path = "my.zarr"
@@ -419,9 +422,10 @@ class DatasetWriterTest(unittest.TestCase, IOCollector):
         ds1.to_zarr(dst_path)
         w = DatasetWriter(dst_path, output_append=True,
                           output_append_dim="t",
-                          output_append_mode=AppendMode.replace)
+                          output_append_mode=AppendMode.replace,
+                          output_consolidated=False)
         w.write_dataset(ds2)
-        ds3 = xr.open_zarr(dst_path)
+        ds3 = xr.open_zarr(dst_path, consolidated=False)
         np.testing.assert_equal(
             np.array(["2001-01-01", "2001-01-02", "2001-01-03",
                       "2001-01-04", "2001-01-05", "2001-01-06"],
@@ -432,26 +436,29 @@ class DatasetWriterTest(unittest.TestCase, IOCollector):
         )
 
     def test_append_overlapping_retain(self):
-        dst_path = "my.zarr"
-        self.add_path(dst_path)
-        ds1, ds2 = new_append_test_datasets(
-            ["2001-01-01", "2001-01-02", "2001-01-03", "2001-01-05"],
-            ["2001-01-03", "2001-01-04", "2001-01-05", "2001-01-06"]
-        )
-        ds1.to_zarr(dst_path)
-        w = DatasetWriter(dst_path, output_append=True,
-                          output_append_dim="t",
-                          output_append_mode=AppendMode.retain)
-        w.write_dataset(ds2)
-        ds3 = xr.open_zarr(dst_path)
-        np.testing.assert_equal(
-            np.array(["2001-01-01", "2001-01-02", "2001-01-03",
-                      "2001-01-04", "2001-01-05", "2001-01-06"],
-                     dtype="datetime64[ns]"), ds3.t.data)
-        np.testing.assert_equal(
-            np.array([0, 0, 0, 1, 0, 1]),
-            ds3.v.isel(x=0, y=0)
-        )
+        for consolidated in False, True:
+            with self.subTest(consolidated=consolidated):
+                dst_path = "my.zarr"
+                self.add_path(dst_path)
+                ds1, ds2 = new_append_test_datasets(
+                    ["2001-01-01", "2001-01-02", "2001-01-03", "2001-01-05"],
+                    ["2001-01-03", "2001-01-04", "2001-01-05", "2001-01-06"]
+                )
+                ds1.to_zarr(dst_path, consolidated=consolidated)
+                w = DatasetWriter(dst_path, output_append=True,
+                                  output_append_dim="t",
+                                  output_append_mode=AppendMode.retain,
+                                  output_consolidated=consolidated)
+                w.write_dataset(ds2)
+                ds3 = xr.open_zarr(dst_path, consolidated=consolidated)
+                np.testing.assert_equal(
+                    np.array(["2001-01-01", "2001-01-02", "2001-01-03",
+                              "2001-01-04", "2001-01-05", "2001-01-06"],
+                             dtype="datetime64[ns]"), ds3.t.data)
+                np.testing.assert_equal(
+                    np.array([0, 0, 0, 1, 0, 1]),
+                    ds3.v.isel(x=0, y=0)
+                )
 
     @classmethod
     def assertTimeSlicesOk(cls, dst_path, src_path_pat, n):
@@ -465,3 +472,5 @@ class DatasetWriterTest(unittest.TestCase, IOCollector):
                     xr.testing.assert_allclose(dst_dataset_slice, src_dataset)
 
     # TODO: add real s3 tests using moto for boto mocking
+
+

